File: client/src/utils/formats.js
export const SUPPORTED_INPUT_FORMATS = new Set([
  'png', 'jpg', 'jpeg', 'webp', 'bmp', 'tiff', 'tif',
  'avif', 'svg', 'ico',
]);

export const ICO_INPUT_FORMATS = new Set([
  'png', 'jpg', 'jpeg', 'webp', 'bmp', 'tiff', 'tif',
  'avif', 'svg',
]);

export const UNSUPPORTED_FORMATS = new Set([
  'gif', 'mp4', 'mov', 'avi', 'mkv', 'webm',
  'mp3', 'wav', 'ogg', 'flac',
  'pdf', 'doc', 'docx', 'xls', 'xlsx', 'ppt', 'txt',
  'heic', 'heif',
]);

export const ALL_OUTPUT_FORMATS = ['PNG', 'JPG', 'WEBP', 'BMP', 'TIFF', 'AVIF', 'SVG'];

/** Get available output formats excluding the source format and ICO */
export function getOutputFormats(sourceFormat) {
  const src = normalizeFormat(sourceFormat);
  return ALL_OUTPUT_FORMATS.filter(f => {
    const n = normalizeFormat(f);
    // exclude source
    if (n === src) return false;
    // jpeg/jpg aliasing
    if ((src === 'jpg' || src === 'jpeg') && (n === 'jpg' || n === 'jpeg')) return false;
    // never show ICO as output
    if (n === 'ico') return false;
    return true;
  });
}

export function normalizeFormat(fmt) {
  if (!fmt) return '';
  const n = fmt.toLowerCase().replace(/^\./, '');
  if (n === 'jpeg') return 'jpg';
  if (n === 'tif')  return 'tiff';
  if (n === 'heif') return 'heic';
  return n;
}

export function getExtFromMime(mime) {
  const map = {
    'image/png':        'png',
    'image/jpeg':       'jpg',
    'image/webp':       'webp',
    'image/bmp':        'bmp',
    'image/tiff':       'tiff',
    'image/avif':       'avif',
    'image/svg+xml':    'svg',
    'image/x-icon':     'ico',
    'application/zip':  'zip',
  };
  return map[mime] || 'bin';
}

export function formatLabel(fmt) {
  if (!fmt) return '';
  const n = normalizeFormat(fmt);
  const labels = {
    jpg: 'JPG', jpeg: 'JPG', png: 'PNG', webp: 'WEBP',
    bmp: 'BMP', tiff: 'TIFF', avif: 'AVIF',
    svg: 'SVG', ico: 'ICO',
  };
  return labels[n] || fmt.toUpperCase();
}

File: py-server/utils/formats.py
"""
Format constants and mappings for Pixora.
"""

SUPPORTED_INPUT_FORMATS = {
    "png", "jpg", "jpeg", "webp", "bmp", "tiff", "tif",
    "avif", "svg", "ico",
}

SUPPORTED_INPUT_MIMES = {
    "image/png",
    "image/jpeg",
    "image/webp",
    "image/bmp",
    "image/tiff",
    "image/avif",
    "image/svg+xml",
    "image/x-icon",
    "image/vnd.microsoft.icon",
}

UNSUPPORTED_FORMATS = {
    "gif", "mp4", "mov", "avi", "mkv", "webm",
    "mp3", "wav", "ogg", "pdf", "doc", "docx",
    "heic", "heif",
}

OUTPUT_FORMATS = ["png", "jpg", "webp", "bmp", "tiff", "avif", "svg"]

ICO_INPUT_FORMATS = {
    "png", "jpg", "jpeg", "webp", "bmp", "tiff", "tif",
    "avif", "svg",
}

FORMAT_TO_PILLOW = {
    "png":  "PNG",
    "jpg":  "JPEG",
    "jpeg": "JPEG",
    "webp": "WEBP",
    "bmp":  "BMP",
    "tiff": "TIFF",
    "tif":  "TIFF",
    "avif": "AVIF",
    "svg":  "SVG",
    "ico":  "ICO",
}

FORMAT_MIME = {
    "png":  "image/png",
    "jpg":  "image/jpeg",
    "jpeg": "image/jpeg",
    "webp": "image/webp",
    "bmp":  "image/bmp",
    "tiff": "image/tiff",
    "tif":  "image/tiff",
    "avif": "image/avif",
    "svg":  "image/svg+xml",
    "ico":  "image/x-icon",
    "zip":  "application/zip",
}

ICO_SIZES = [16, 32, 48, 64, 128, 256]


def normalize_format(fmt: str) -> str:
    """Normalize format string to lowercase without dots."""
    return fmt.lower().lstrip(".")


def get_mime(fmt: str) -> str:
    fmt = normalize_format(fmt)
    return FORMAT_MIME.get(fmt, "application/octet-stream")


def get_pillow_format(fmt: str) -> str:
    fmt = normalize_format(fmt)
    return FORMAT_TO_PILLOW.get(fmt, fmt.upper())